Fix type annotations incorrectly highlighted as modules.

Wilfred opened the pull request:
Previously, we were always treating `::` as a module, but Rust
allows type annotations using `::` e.g.

    parse::<i32>();

This also changes module highlighting so that only the module name is
highlighted, excluding the `::`. This makes rust-mode consistent with
other Emacs modes, such as c++-mode and ruby-mode.

Before:

![rust_before](https://cloud.githubusercontent.com/assets/70800/12070879/af06707a-b086-11e5-97ab-64b9eb1b8363.png)

After:

![rust_after](https://cloud.githubusercontent.com/assets/70800/12070882/c0932fb8-b086-11e5-9097-6fa8e4bfa677.png)

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

Wilfred commented:
@nikomatsakis I've pushed a fix for highlighting of `foo:: <bar>` and added a matching test.

I've also refactored the while loop because I thought it was a little hard to read.

Let me know what you think or if you'd like any other changes. :)

nikomatsakis commented:
@Wilfred seems good to me! thanks.
